src/Enzyme.jl

```diff
@@ -679,15 +747,17 @@ result, ∂v, ∂A
 
     @assert ReturnShadow
     TT = Tuple{args...}
-   
+
     primal_tt = Tuple{map(eltype, args)...}
     world = codegen_world_age(eltype(FA), primal_tt)
 
     # TODO this assumes that the thunk here has the correct parent/etc things for getting the right cuda instructions -> same caching behavior
+    # XXX: Calling thunk from autodiff_deferred_thunk is invalid.
     nondef = Enzyme.Compiler.thunk(Val(world), FA, A, TT, #=Split=# Val(API.DEM_ReverseModeGradient), Val(width), ModifiedBetween, #=ReturnPrimal=#Val(ReturnPrimal), #=ShadowInit=#Val(false), RABI)
     TapeType = EnzymeRules.tape_type(nondef[1])
     A2 = Compiler.return_type(typeof(nondef[1]))
```

Comment:
just have this call autodiff_deferred_thunk(TapeType, ...)

Reply:
That would repeat the boilerplate code before the `thunk`, plus still return and do the `asserts`. So not sure that helps anything. @vchuravy mentioned that we should maybe get rid of the tape typeless call?